src/plugins/intel_npu/src/al/include/intel_npu/network_metadata.hpp

```diff
@@ -0,0 +1,127 @@
+     */
+    std::string nameFromCompiler;
+
+    ov::element::Type precision;
+
+    ov::PartialShape shapeFromCompiler;
+
+    /**
+     * @brief If set to "true", the current object describes a buffer which may be used for altering a state tensor.
+     * @details This flag is set if the compiler prefixed the name using a "read value" prefix. The state input and
+     * state output descriptors are also tied using the "relatedDescriptorIndex" attribute.
+     */
+    bool isStateInput = false;
+
+    /**
+     * @brief If set to "true", the current object describes a buffer which reflects the value of a state tensor.
+     * @details This flag is set if the compiler prefixed the name using an "assign" prefix. The state input and
+     * state output descriptors are also tied using the "relatedDescriptorIndex" attribute.
+     */
+    bool isStateOutput = false;
+
+    /**
+     * @brief If set to "true", the buffer of the tensor described here contains as value the shape of the referenced
+     * tensor.
+     * @details This flag is set if the compiler prefixed the name using a "shape" prefix.
+     *
+     * The referenced tensor bears the same name ("nameFromCompiler"), but its "isShapeTensor" value is set to
+     * "false". The two descriptors are also tied using the "relatedDescriptorIndex" attribute.
+     */
+    bool isShapeTensor = false;
+
+    /**
+     * @brief Points towards a related descriptor.
+     * @details The related descriptors are defined by (state input, state output) or (dynamic tensor, shape tensor)
+     * pairs.
+     */
+    std::optional<size_t> relatedDescriptorIndex;
+
+    /**
+     * @brief The friendly name of the node extracted from the IR model.
+     * @details In some cases, this field is required for constructing a dummy model which uses the same input/output
+     * metadata as the original IR model.
+     *
+     * This field may be empty if the I/O entry is not found in the original IR model (i.e. the entry was added by the
+     * compiler).
+     */
+    std::string nodeFriendlyName;
+
+    /**
+     * @brief The names of the output tensors extracted from the IR model.
+     * @details In some cases, this field is required for constructing a dummy model which uses the same input/output
+     * metadata as the original IR model.
+     *
+     * This field may be empty if the I/O entry is not found in the original IR model (i.e. the entry was added by the
+     * compiler).
+     */
+    std::unordered_set<std::string> outputTensorNames;
+
+    /**
+     * @brief The shape extracted from the IR model.
+     * @details The values may differ from the ones found in "shapeFromCompiler" if batching is to be handled by the
+     * plugin.
+     *
+     * This field may be empty if the I/O entry is not found in the original IR model (i.e. the entry was added
+     * by the compiler).
+     */
+    std::optional<ov::PartialShape> shapeFromIRModel = std::nullopt;
+};
+
+struct NetworkMetadata final {
```

Comment:
Not sure if we shall change NetworkMetadata to class, then can inherit and produce Metadata with private member for different compiler and not break plugin interface.

Reply:
I could imagine that we name `NetworkMetadata` an implementation detail and would not expose it in `NPU AL API`, but it is more like an improvement, can be done in a separate PR.